libraries/tickfs/SPEC.md

```diff
@@ -0,0 +1,494 @@
+
+Where ONE, TWO and FOUR have been marked as invalid the entire region will still
+be classified as full. It will only not be full when all four objects are marked
+for invalid.
+
+It would be possible to have a reserved region and then free up the region with
+the following steps:
+ 1. Move the region to the reserved region
+ 1. Erase the above region
+ 1. Move the valid objects from the reserved region to the original region,
+    after being defraged in memory
+ 1. Erase the reserved region
+
+The above steps would ensure we don't loose data on a power loss, as long as
+we could ensure we could continue the operation after regaining power.
+
+There are a few problems with this approach though.
+
+In order to maintain "garbage collecting" state to restore after a power loss
+we would need some sore of super block. This would then break the requirement
+on wear leveling as whichever region that block is in will be written and
+erased more often the others.
+
+The other problem is the reserved region where we would move data to. This
+would cost space as an entire region would be dedicated to garbage collecting.
+The garbage collecting region would also have more erase and writes performed
+on it breaking the wear leveling requirement.
+
+### Somewhat high storage overhead
+
+The storage overhead is somewhat high for TickFS. This is mostly due to the
+two 64-bit hashes that are stored with every object.
+
+The 64-bit value is determined by the output of the standard Rust
+`core::hash::Hasher` trait.
+
+For the hash of the key a 64-bit value can be justified by the lack of
+collision avoidance in the implementation. If two keys have the same
+hash the second key will be dropped. In this case a 64-bit hash should
+hopefully make that occurrence very unlikely. Also by following the standard
+Rust `core::hash::Hasher` trait the user is free to implement any standard
+Hasher of their choosing. Some systems will even be able to offload the
+hash to hardware.
+
+As for the check-sum. It was seriously considered to change this to a
+32-bit or even 16-bit CRC. This would reduce the overhead for each object
+by 4 or 6 bytes. Although a CRC is relatively easy to implement, it can still
+be time consuming. For this reason lots of embedded hardware will include a
+CRC offload engine to offload the calculation. By implementing a CRC in this
+library I would remove the option of allowing a offload engine. As there
+doesn't seem to be a generic CRC Rust trait to take advantage of the decision
+was made to reuse the `core::hash::Hasher` trait for the checksum. This costs
+space, but hopefully reduces implementation burden and allows for more hardware
+offloading.
+
+### Memory usage
+
+At a minimum TickFS requires a `PageSize` amount of free memory to store the
+buffer obtained when reading flash. This could be reduced by instead performing
+multiple reads of a smaller fixed size.
+
+### Synchronous
+
+TickFS is fully synchronous and every call is blocking.
```

Comment:
Do you have an idea as to how you would implement this? I'm guessing by using a new flash driver?

Reply:
I think no matter what a new flash driver would need to be implemented. The current Tock interfaces are very Tock specific and I was trying to make this generic.

There are two options (or both):
 1. Update TickFS to also support asynchronous operations, we can use the callback style method used in Tock. This would take some work, but isn't impossible. Then we would just need a small translation layer to convert between Tock and TickFS.
 1. Add a new flash interface using synchronous calls.

Reply:
The flash HIL is read/write/erase page with callbacks. I don't see how that is overly Tock specific. There may have to be a shim layer to mostly do address conversion/restriction, but that would be a standard shim that any board with a flash driver could use.

The other benefit to using an asynchronous interface is that this FS would work on an external flash chip (like the nrf52840dk has) as well.

Reply:
It only requires unsafe code to construct a static buffer that is initially passed to the capsule using the filesystem, so it is at least contained within `main.rs`. But I see that if not for the requirement of async operations, an anonymous lifetime would work, which is much nicer as it would allow clients of the filesystem to use stack allocated buffers for filesystem reads/writes

Reply:
This needs to be asynchronous: we talked about this in a call. While on-chip flash can behave synchronously, that's only true if you're writing to the bank you are executing out of. If you have off-chip flash (e.g., over a SPI) you do not want to put a blocking interface on top of it.

Reply:
The interface in Tock will be async, see https://github.com/tock/tock/pull/2205 for how this is done.